File: androidApp/src/main/java/com/dev/kmmwallet/androidApp/viewmodel/HomeViewModel.kt
package com.dev.kmmwallet.androidApp.viewmodel

import com.dev.kmmwallet.shared.viewmodel.event.home.HomeScreenState
import com.dev.kmmwallet.shared.viewmodel.event.home.initHomeScreen
import com.dev.kmmwallet.shared.viewmodel.event.home.logoutUser
import java.util.*

class HomeViewModel : BaseViewModel() {

    fun initHomePage(updateHomeUi: (homeScreenState: HomeScreenState) -> Unit) {
        sharedViewModel.stateProvider.initHomeScreen {
            it.greeting = greetUser()
            updateHomeUi(it)
        }
    }

    private fun greetUser(): String {
        val c: Calendar = Calendar.getInstance()
        val timeOfDay: Int = c.get(Calendar.HOUR_OF_DAY)
        if (timeOfDay in 0..11) {
            return "Good Morning"
        } else if (timeOfDay in 12..15) {
            return "Good Afternoon"
        } else if (timeOfDay in 16..20) {
            return "Good Evening"
        } else if (timeOfDay in 21..23) {
            return "Good Night"
        }
        return ""
    }

    fun logoutUser(onLogout: (logoutState: HomeScreenState) -> Unit) {
        sharedViewModel.stateProvider.logoutUser(){
            onLogout(it)
        }
    }

}

File: iosApp/iosApp/ViewModel/HomeViewModel.swift
//
//  HomeViewModel.swift
//  iosApp
//
//  Created by Chinthaka on 05/04/2021.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import shared

class HomeViewModel {
    private let sharedViewModel: SharedViewModel = SharedViewModel()
    private func greetUser() -> String{
        
        let date = NSDate()
        let calendar = NSCalendar.current
        let currentHour = calendar.component(.hour, from: date as Date)
        let hourInt = Int(currentHour.description)!
        
        if hourInt >= 12 && hourInt <= 16 {
            return  "Good Afternoon"
        }
        else if (hourInt >= 7 && hourInt <= 12) || (hourInt >= 0 && hourInt <= 7)  {
            return "Good Morning"
        }
        else if hourInt >= 16 && hourInt <= 20 {
            return "Good Evening"
        }
        else if hourInt >= 20 && hourInt <= 24 {
            return "Good Night"
        }
        return ""
    }
    
    func initHome(completion: @escaping (HomeScreenState)->()){
        self.sharedViewModel.stateProvider.doInitHomeScreen { (state: HomeScreenState) in
            state.greeting = self.greetUser()
            completion(state)
        }
    }
}
